src/node_jsvmapi.cc

```diff
@@ -465,22 +465,109 @@ namespace v8impl {
 
 }  // end of namespace v8impl
 
-// We define a macro here because, with full error handling, we may want to
-// perform additional actions at the top of our public APIs, such as bail,
-// returning an error code at the very top of each public API in the case that
-// there's already an exception pending in v8impl::TryCatch::lastException
-#define NAPI_PREAMBLE(e) \
-  v8impl::TryCatch tryCatch(v8impl::V8IsolateFromJsEnv((e)))
+#define RETURN_STATUS_IF_FAIL(condition, status)                        \
+  do {                                                                  \
+    if (!(condition)) {                                                 \
+      return napi_set_last_error_code((status));                        \
+    }                                                                   \
+  } while(0)
+
+#define CHECK_ARG(arg)                                                  \
+  RETURN_STATUS_IF_FAIL((arg), napi_invalid_arg)
+
+#define CHECK_MAYBE_EMPTY(maybe, status)                                \
+  RETURN_STATUS_IF_FAIL(!((maybe).IsEmpty()), (status))
+
+#define CHECK_MAYBE_NOTHING(maybe, status)                              \
+  RETURN_STATUS_IF_FAIL(!((maybe).IsNothing()), (status))
+
+#define NAPI_PREAMBLE(e)                                                \
+  do {                                                                  \
+    RETURN_STATUS_IF_FAIL(v8impl::TryCatch::lastException().IsEmpty(),  \
+      napi_pending_exception);                                          \
+    napi_clear_last_error_code();                                       \
+    v8impl::TryCatch tryCatch(v8impl::V8IsolateFromJsEnv((e)));         \
+  } while(0)
+
+#define CHECK_TO_OBJECT(result, o)                                      \
+  do {                                                                  \
+    auto obj_maybe =                                                    \
+      v8impl::V8LocalValueFromJsValue((o))->ToObject(context);          \
+    CHECK_MAYBE_EMPTY(obj_maybe, napi_object_expected);                 \
+    result = obj_maybe.ToLocalChecked();                                \
+  } while(0)
+
+#define CHECK_NEW_FROM_UTF8_LEN(result, str, len)                       \
+  do {                                                                  \
+    auto name_maybe = v8::String::NewFromUtf8(isolate, (str),           \
+      v8::NewStringType::kInternalized, len);                           \
+    CHECK_MAYBE_EMPTY(name_maybe, napi_generic_failure);                \
+    result = name_maybe.ToLocalChecked();                               \
+  } while(0)
+
+#define CHECK_NEW_FROM_UTF8(result, str)                                \
+    CHECK_NEW_FROM_UTF8_LEN((result), (str), -1)
+
+   // Static last error returned from napi_get_last_error_info
+napi_extended_error_info static_last_error;
+
+// Warning: Keep in-sync with napi_status enum
+const char* error_messages[] = {
+  NULL,
```

Comment:
I think it shouldn't be an issue as long as it's in the .cc file. I suppose we do want to keep the .h file C-compatible though.